### DISCUSSION TITLE
Not all many to many releations are getting saved

### DISCUSSION THREAD

**Discussion by Marvin-S** [CodeDredd/pinia-orm#1105] [Q&A] [answered]
Hello everybody,

I'm experiencing a strange behavior with the many to many relationship in my application.
I've worked through the [Documentation](https://pinia-orm.codedredd.de/guide/relationships/many-to-many#many-to-many) and created 3 models:
1. OutsourcingPartner
2. BillingGroup
3. OutsourcingPartnerBillingGroup

Obviously the outsourcing partner and the billing group have a many to many relationship.

I recieve data from an API that returns the outsourcing partners with all their billing groups, but only some og the relations are getting saved.

My models are looking like this:
```typescript
class OutsourcingPartner extends Model {
  static entity = 'outsourcingPartners';

  @Uid()
  declare id: string;

  @Str('')
  declare name: string;

  @BelongsToMany(
    () => BillingGroup,
    () => OutsourcingPartnerBillingGroup,
    'outsourcingPartner_id',
    'billingGroup_id'
  )
  declare billingGroups: BillingGroup[];
}

class BillingGroup extends Model {
  static entity = 'billingGroups';

  @Uid()
  declare id: string;

  @Str('')
  declare name: string;
}


class OutsourcingPartnerBillingGroup extends Model {
  static entity = 'outsourcingPartnerBillingGroups';

  static primaryKey = ['outsourcingPartner_id', 'billingGroup_id'];

  @Uid()
  declare outsourcingPartner_id: string;

  @Uid()
  declare billingGroup_id: string;
}
```

The data that I recieve from my API and pass to the `save()` function is the following:
```json
[
    {
       "id":"36a11c85380b4afb994710ea63f99964",
       "name":"Luke Skywalker",
       "billingGroups":[
          {
             "id":"c8d8db2b1f4f4a069fa640c9c8cbe489",
             "name":"The Jedis",
             "__typename":"BillingGroup"
          }
       ],
       "__typename":"OutsourcingPartner"
    },
    {
       "id":"658b19b76a0f494590c21a7c27173d8e",
       "name":"Darth Vader",
       "billingGroups":[
          {
             "id":"95a87597ef124bbf9907f8f6074f8e94",
             "name":"Dark Forces",
             "__typename":"BillingGroup"
          }
       ],
       "__typename":"OutsourcingPartner"
    },
    {
       "id":"fabaecbb6ac045bb9b582244b889d309",
       "name":"Yoda",
       "billingGroups":[
          {
             "id":"4dedf55e5a8a4ce5affb1bc65b65de09",
             "name":"Jedi Council",
             "__typename":"BillingGroup"
          },
          {
             "id":"c8d8db2b1f4f4a069fa640c9c8cbe489",
             "name":"The Jedis",
             "__typename":"BillingGroup"
          }
       ],
       "__typename":"OutsourcingPartner"
    },
    {
       "id":"0284241ba3ed487eb8d12c0e86161fc0",
       "name":"Clone Trooper 0815",
       "billingGroups":[
          {
             "id":"95a87597ef124bbf9907f8f6074f8e94",
             "name":"Dark Forces",
             "__typename":"BillingGroup"
          }
       ],
       "__typename":"OutsourcingPartner"
    }
 ]
```

The result from the `save()` function is the following:
```json
[
    {
        "id": "36a11c85380b4afb994710ea63f99964",
        "name": "Luke Skywalker",
        "billingGroups": [
            {
                "id": "c8d8db2b1f4f4a069fa640c9c8cbe489",
                "name": "The Jedis",
                "pivot_outsourcingPartnerBillingGroups": null
            }
        ]
    },
    {
        "id": "658b19b76a0f494590c21a7c27173d8e",
        "name": "Darth Vader",
        "billingGroups": [
            {
                "id": "95a87597ef124bbf9907f8f6074f8e94",
                "name": "Dark Forces",
                "pivot_outsourcingPartnerBillingGroups": null
            }
        ]
    },
    {
        "id": "fabaecbb6ac045bb9b582244b889d309",
        "name": "Yoda",
        "billingGroups": [
            {
                "id": "4dedf55e5a8a4ce5affb1bc65b65de09",
                "name": "Jedi Council",
                "pivot_outsourcingPartnerBillingGroups": {
                    "outsourcingPartner_id": "fabaecbb6ac045bb9b582244b889d309",
                    "billingGroup_id": "4dedf55e5a8a4ce5affb1bc65b65de09"
                }
            },
            {
                "id": "c8d8db2b1f4f4a069fa640c9c8cbe489",
                "name": "The Jedis",
                "pivot_outsourcingPartnerBillingGroups": null
            }
        ]
    },
    {
        "id": "0284241ba3ed487eb8d12c0e86161fc0",
        "name": "Clone Trooper 0815",
        "billingGroups": [
            {
                "id": "95a87597ef124bbf9907f8f6074f8e94",
                "name": "Dark Forces",
                "pivot_outsourcingPartnerBillingGroups": null
            }
        ]
    }
]
```

And the state store is looking like this after the call to the `save()` function:
```json
{
  "outsourcingPartners": {
    "data": {
      "36a11c85380b4afb994710ea63f99964": {
        "id": "36a11c85380b4afb994710ea63f99964",
        "name": "Luke Skywalker"
      },
      "658b19b76a0f494590c21a7c27173d8e": {
        "id": "658b19b76a0f494590c21a7c27173d8e",
        "name": "Darth Vader"
      },
      "fabaecbb6ac045bb9b582244b889d309": {
        "id": "fabaecbb6ac045bb9b582244b889d309",
        "name": "Yoda"
      },
      "0284241ba3ed487eb8d12c0e86161fc0": {
        "id": "0284241ba3ed487eb8d12c0e86161fc0",
        "name": "Clone Trooper 0815"
      }
    }
  },
  "outsourcingPartnerBillingGroups": {
    "data": {
      "[\"36a11c85380b4afb994710ea63f99964\",\"c8d8db2b1f4f4a069fa640c9c8cbe489\"]": {
        "outsourcingPartner_id": "36a11c85380b4afb994710ea63f99964",
        "billingGroup_id": "c8d8db2b1f4f4a069fa640c9c8cbe489"
      },
      "[\"658b19b76a0f494590c21a7c27173d8e\",\"95a87597ef124bbf9907f8f6074f8e94\"]": {
        "outsourcingPartner_id": "658b19b76a0f494590c21a7c27173d8e",
        "billingGroup_id": "95a87597ef124bbf9907f8f6074f8e94"
      },
      "[\"fabaecbb6ac045bb9b582244b889d309\",\"4dedf55e5a8a4ce5affb1bc65b65de09\"]": {
        "outsourcingPartner_id": "fabaecbb6ac045bb9b582244b889d309",
        "billingGroup_id": "4dedf55e5a8a4ce5affb1bc65b65de09"
      }
    }
  },
  "billingGroups": {
    "data": {
      "c8d8db2b1f4f4a069fa640c9c8cbe489": {
        "id": "c8d8db2b1f4f4a069fa640c9c8cbe489",
        "name": "The Jedis"
      },
      "95a87597ef124bbf9907f8f6074f8e94": {
        "id": "95a87597ef124bbf9907f8f6074f8e94",
        "name": "Dark Forces"
      },
      "4dedf55e5a8a4ce5affb1bc65b65de09": {
        "id": "4dedf55e5a8a4ce5affb1bc65b65de09",
        "name": "Jedi Council"
      }
    }
  }
}
```

So as you can see, only three relations are getting stored, but I have five in total. And also if I query the model from the repo, I only recieve three relations.

I'm pretty sure there is some kind of misconfiguration from my side, but I don't know where.

Any help is really appreciated.

**Comment by CodeDredd** [comment, marked as answer]
Hey @Marvin-S ,

i tried your code and got the right result:

````ts
{
  outsourcingPartnerBillingGroups: {
    data: {
      '["36a11c85380b4afb994710ea63f99964","c8d8db2b1f4f4a069fa640c9c8cbe489"]': [Object],
      '["658b19b76a0f494590c21a7c27173d8e","95a87597ef124bbf9907f8f6074f8e94"]': [Object],
      '["fabaecbb6ac045bb9b582244b889d309","4dedf55e5a8a4ce5affb1bc65b65de09"]': [Object],
      '["fabaecbb6ac045bb9b582244b889d309","c8d8db2b1f4f4a069fa640c9c8cbe489"]': [Object],
      '["0284241ba3ed487eb8d12c0e86161fc0","95a87597ef124bbf9907f8f6074f8e94"]': [Object]
    }
  },
  billingGroups: {
    data: {
      c8d8db2b1f4f4a069fa640c9c8cbe489: [Object],
      '95a87597ef124bbf9907f8f6074f8e94': [Object],
      '4dedf55e5a8a4ce5affb1bc65b65de09': [Object]
    }
  },
  outsourcingPartners: {
    data: {
      '36a11c85380b4afb994710ea63f99964': [Object],
      '658b19b76a0f494590c21a7c27173d8e': [Object],
      fabaecbb6ac045bb9b582244b889d309: [Object],
      '0284241ba3ed487eb8d12c0e86161fc0': [Object]
    }
  }
}
````

If you can give me a working reproduction then i can help you. 🙏  Right now it seems some important information is missing.

**Reply by Marvin-S** [reply]
Hi @CodeDredd,

i digged a bit deeper inside this problem and also created a [Codesandbox](https://codesandbox.io/p/sandbox/pinia-orm-with-apollo-wiyhrr) for this.
The application uses quasar, apollo, pinia and pinia-orm. I mocked the GraphQL API with Mockend (I needed to change the test data slightly to get the mock server running).

I found out that the strange behavior I described above only occurs with the Apollo DevTools activated. If I deactivate them, the behavior is totally different, but still not working.

# Behavior with Apollo DevTools
Here is the behaviour that I experience with activated Apollo DevTools:
![image](https://user-images.githubusercontent.com/52828682/227192906-dc4d1a9e-fb07-4d8c-b335-32e3674d33ae.png)
What I think is pretty strange is, that it logs the `pivot_...` fields in the API result object although it does not contain this data.
But the result object does only contain null values on the `pivot_...` fields.

# Behavior with deactivated Apollo DevTools
What is even more interesting is, that if I open the exact same site with deactivated Apollo DevTools:
I recieve an error from the save method that states, that it cannot add the `pivot_...` fields.
![image](https://user-images.githubusercontent.com/52828682/227195054-98a20ca1-09bb-493c-b8e5-9142b6db3712.png)
So as you can see, the API response does not have the `pivot_...` fields but I recieve an error from the `save()` method.
I've found out, that if I hardcode the test data I can call the `save()` method without a problem. So I guess there are some complications with the Apollo result object.

The Codesandbox contains both:
1. Saving the hardcoded test data
2. Saving the actual API response
So if you take a look at the state store, please keep in mind, that the data gets saved with the hardcoded data.

**Reply by Marvin-S** [reply]
Hey @CodeDredd,

do you have an idea how to get around this complications with the Apollo result object? I'm a bit stuck on this and maybe there is an easy trick for that, that I don't know about.

**Reply by CodeDredd** [reply]
i just read your comments and thinking about it.

**Reply by CodeDredd** [reply]
ok. I found the problem. 
First the result direct from the save method returns the result with the helper variable to create a correct pivot relation.
As you right guessed the problem lies with the apollo response. It return an non extensible object: https://stackoverflow.com/questions/45122831/why-does-apollo-graphql-return-a-non-extensible-object

The workaround for this is to deepclone the result of the graphql object before it is saved.

i also see that the quasar apollo extension is still using "apollo-client": "^2.6.8". Maybe the issue is also resolved if you just upgrade the apollo client to 3.7.10.

**Reply by Marvin-S** [reply]
Thank you very much for your help!

I've tried to use the lodash `cloneDeep` function. The codesandbox was already using Apollo Client 3.7.10.
Now I have the same behavior as in my original post:
Only a few releations are getting saved. If I use the `JSON.parse(JSON.stringify(...))` method, then it works, but this is not a good solution in my opinion, because of potential data loss.

I've updated the [codesandbox example](https://codesandbox.io/p/sandbox/pinia-orm-with-apollo-wiyhrr?file=%2Fquasar-project%2Fsrc%2Frepositories%2FOutsourcingPartnerRepository.ts) for reproduction and also checked if the Apollo DevTools were somehow activated again :D

**Reply by CodeDredd** [reply]
if i open the codesandbox example i get a 500 error:
`
-1 - Error: (500) - PATH: /beta/sandboxes/branches/wiyhrr/instance - DATA: {"code":"UNKNOWN_ERROR","error":"UNKNOWN_ERROR - authorize error: pitcher-host-container didn't respond in 5s","message":"authorize error: pitcher-host-container didn't respond in 5s"} - ERROR: Error: (500) UNKNOWN_ERROR - authorize error: pitcher-host-container didn't respond in 5s
`

**Reply by Marvin-S** [reply]
Should work now, I just restarted the instance.

**Reply by CodeDredd** [reply]
i am confused....i can't see an error: The api response is this:
````json
[
    {
        "__typename": "Outsourcingpartner",
        "id": 1,
        "name": "Luke Skywalker",
        "billingGroups": [
            {
                "__typename": "Billinggroup",
                "id": 1,
                "name": "Dark Forces"
            },
            {
                "__typename": "Billinggroup",
                "id": 2,
                "name": "Jedi Council"
            },
            {
                "__typename": "Billinggroup",
                "id": 3,
                "name": "The Jedis"
            }
        ]
    },
    {
        "__typename": "Outsourcingpartner",
        "id": 2,
        "name": "Darth Vader",
        "billingGroups": [
            {
                "__typename": "Billinggroup",
                "id": 1,
                "name": "Dark Forces"
            },
            {
                "__typename": "Billinggroup",
                "id": 2,
                "name": "Jedi Council"
            },
            {
                "__typename": "Billinggroup",
                "id": 3,
                "name": "The Jedis"
            }
        ]
    },
    {
        "__typename": "Outsourcingpartner",
        "id": 3,
        "name": "Yoda",
        "billingGroups": [
            {
                "__typename": "Billinggroup",
                "id": 1,
                "name": "Dark Forces"
            },
            {
                "__typename": "Billinggroup",
                "id": 2,
                "name": "Jedi Council"
            },
            {
                "__typename": "Billinggroup",
                "id": 3,
                "name": "The Jedis"
            }
        ]
    },
    {
        "__typename": "Outsourcingpartner",
        "id": 4,
        "name": "Clone Trooper 0815",
        "billingGroups": [
            {
                "__typename": "Billinggroup",
                "id": 1,
                "name": "Dark Forces"
            },
            {
                "__typename": "Billinggroup",
                "id": 2,
                "name": "Jedi Council"
            },
            {
                "__typename": "Billinggroup",
                "id": 3,
                "name": "The Jedis"
            }
        ]
    }
]
````

And that is exactly what is saved:
````json
[
    {
        "id": 1,
        "name": "Luke Skywalker",
        "billingGroups": [
            {
                "id": 1,
                "name": "Dark Forces",
                "pivot_outsourcingPartnerBillingGroups": null
            },
            {
                "id": 2,
                "name": "Jedi Council",
                "pivot_outsourcingPartnerBillingGroups": null
            },
            {
                "id": 3,
                "name": "The Jedis",
                "pivot_outsourcingPartnerBillingGroups": null
            }
        ]
    },
    {
        "id": 2,
        "name": "Darth Vader",
        "billingGroups": [
            {
                "id": 1,
                "name": "Dark Forces",
                "pivot_outsourcingPartnerBillingGroups": null
            },
            {
                "id": 2,
                "name": "Jedi Council",
                "pivot_outsourcingPartnerBillingGroups": null
            },
            {
                "id": 3,
                "name": "The Jedis",
                "pivot_outsourcingPartnerBillingGroups": null
            }
        ]
    },
    {
        "id": 3,
        "name": "Yoda",
        "billingGroups": [
            {
                "id": 1,
                "name": "Dark Forces",
                "pivot_outsourcingPartnerBillingGroups": null
            },
            {
                "id": 2,
                "name": "Jedi Council",
                "pivot_outsourcingPartnerBillingGroups": null
            },
            {
                "id": 3,
                "name": "The Jedis",
                "pivot_outsourcingPartnerBillingGroups": null
            }
        ]
    },
    {
        "id": 4,
        "name": "Clone Trooper 0815",
        "billingGroups": [
            {
                "id": 1,
                "name": "Dark Forces",
                "pivot_outsourcingPartnerBillingGroups": null
            },
            {
                "id": 2,
                "name": "Jedi Council",
                "pivot_outsourcingPartnerBillingGroups": null
            },
            {
                "id": 3,
                "name": "The Jedis",
                "pivot_outsourcingPartnerBillingGroups": null
            }
        ]
    }
]
````

**Reply by Marvin-S** [reply]
Exactly! But if you take a look at the state store you will only see 3 entries for the relationships instead of 12:
![image](https://user-images.githubusercontent.com/52828682/228137990-0d932413-8e45-4a70-a1b6-1ad8ca7cd64b.png)

And I don't know if it's right that the `pivot_...` fields are null after saving.

**Reply by CodeDredd** [reply]
I have to admit....i have no clue what is happning....trying to make a unit test of it. Takes some time because i need graphql for it.

**Reply by Marvin-S** [reply]
Thank you for your help!
If I can help you with that, just tell me, I can also give you access to my mockend graphql endpoint if you want to :)

**Reply by CodeDredd** [reply]
@Marvin-S Help is welcome. I started a new pr. Right now it doesn't work. Maybe you get it running #1147....
lets see who is first making the test work ^^.

**Reply by Marvin-S** [reply]
Ok I will take a look at it tomorrow.

**Reply by CodeDredd** [reply]
ok the query works now...but the test works...that is annoying ^^.
Seems to be an issue the the cache provider i guess.

**Reply by CodeDredd** [reply]
Well as a workaround instead of using "cloneDeep" you can use "JSON.parse(JSON.stringfy())"....as you already mentioned that is no ideal.
But this bug seems to be a really hard one.

**Reply by Marvin-S** [reply]
Hi Gregor,

I think I got a little further on that.
I created an own test that queries my mockend GraphQL endpoint using Apollo (I know this will not be merged in the future, but it helped me to figure out what is happening here).

I found some interesting stuff during debugging. As far as I can see, the problem lies between the Schema.ts inside the idAttribute callback function and the Normalizr code.

The data gets normalized recursively. So it loops through every Outsourcing Partner, each of its Billing Groups and each of their pivot fields. On the first run of the Outsourcing Partners everything seems to work fine. But on the second run it begins to act weird. Normalizr has a `normalize` function that does some checks and calls the `visit` function recursively:
```javascript
normalize(input, parent, key, visit, addEntity, visitedEntities) {
    const id = this.getId(input, parent, key);
    const entityType = this.key;
    if (!(entityType in visitedEntities)) {
      visitedEntities[entityType] = {};
    }
    if (!(id in visitedEntities[entityType])) {
      visitedEntities[entityType][id] = [];
    }
    if (visitedEntities[entityType][id].includes(input)) {
      return id;
    }
    visitedEntities[entityType][id].push(input);
    const processedEntity = this._processStrategy(input, parent, key);
    Object.keys(this.schema).forEach((key2) => {
      if (processedEntity.hasOwnProperty(key2) && typeof processedEntity[key2] === "object") {
        const schema = this.schema[key2];
        const resolvedSchema = typeof schema === "function" ? schema(input) : schema;
        processedEntity[key2] = visit(
          processedEntity[key2],
          processedEntity,
          key2,
          resolvedSchema,
          addEntity,
          visitedEntities
        );
      }
    });
    addEntity(this, processedEntity, input, parent, key);
    return id;
  }
```

In this piece of code inside this function, it checks if the entity already got visited and just returns if it has:
```javascript
if (visitedEntities[entityType][id].includes(input)) {
      return id;
 }
```

Surprisingly after the first run (to be precise on the second Outsourcing Partner and its first Billing Group), Normalizr thinks, that it already visited the entity and skips further processing.

So I tried to figure out why it thinks that this entity already got visited, because it is not only checking if the Billing Group Id is inside the `visitedEntities` object but if the whole Billing Group with its `pivot_outsourcingPartnerBillingGroups` is inside the `visitedEntities`. Normally that shouldn't be the case because the `pivot_outsourcingPartnerBillingGroups` should contain other Ids on the second run, but as the program enters this `normalize` function, the old Ids from the first run are inside the object. Here is a snapshot of the whole Outsourcing Partner that gets processed on the third run:
```javascript
{
  __typename: "Outsourcingpartner",
  id: 3,
  name: "Yoda",
  billingGroups: [
    {
      __typename: "Billinggroup",
      id: 1,
      name: "Dark Forces",
      pivot_outsourcingPartnerBillingGroups: {
        outsourcingPartner_id: 2,
        billingGroup_id: 1,
      },
    },
    {
      __typename: "Billinggroup",
      id: 2,
      name: "Jedi Council",
      pivot_outsourcingPartnerBillingGroups: {
        outsourcingPartner_id: 2,
        billingGroup_id: 2,
      },
    },
    {
      __typename: "Billinggroup",
      id: 3,
      name: "The Jedis",
      pivot_outsourcingPartnerBillingGroups: {
        outsourcingPartner_id: 2,
        billingGroup_id: 3,
      },
    },
  ],
}
```

As you can see, the Outsourcing Partner itself has the Id 3, but all its `billingGroups.pivot_outsourcingPartnerBillingGroups` have the `outsourcingPartner_id` of 2.

So what happens with that? Each Billing Group will be normalized through the `normalize` function. On the first line it tries to get the id:
```javascript
const id = this.getId(input, parent, key);
```
This call maps to the `idAttribute` callback function inside the Schema.ts file of pinia-orm:
```typescript
private idAttribute(
    model: Model,
    parent: Model,
  ): Normalizr.StrategyFunction<string> {
    // We'll first check if the model contains any uid attributes. If so, we
    // generate the uids during the normalization process, so we'll keep that
    // check result here. This way, we can use this result while processing each
    // record, instead of looping through the model fields each time.
    const uidFields = this.getUidPrimaryKeyPairs(model)

    return (record, parentRecord, key) => {
      // If the `key` is not `null`, that means this record is a nested
      // relationship of the parent model. In this case, we'll attach any
      // missing foreign keys to the record first.
      if (key !== null)
        (parent.$fields()[key] as Relation).attach(parentRecord, record)

      // Next, we'll generate any missing primary key fields defined as
      // uid field.
      for (const key in uidFields) {
        if (isNullish(record[key]))
          record[key] = uidFields[key].setKey(key).make(record[key])
      }

      // Finally, obtain the index id, attach it to the current record at the
      // special `__id` key. The `__id` key is used when we try to retrieve
      // the models via the `revive` method using the data that is currently
      // being normalized.
      const id = model.$getIndexId(record)

      return id
    }
  }
```
Here it first tries to attach the foreign key. By doing so it changes the `pivot_outsourcingPartnerBillingGroups` object with the updated IDs (so for Outsourcing Parter 3, it will change the `outsourcingPartner_id` field to 3). On the first glance that seemed fine for me. But for some reason, that I do not unterstand is that is also changes the value inside the `visitedEntities` of Normalizr. So at this point Normalizr thinks, that it already visited this entity and will not process it any further.

What's also interesting is, that after processing the first Outsourcing Partner, all the others Outsourcing Partners Billing Groups have the `pivot_outsourcingPartnerBillingGroups` set. Of course for the other Outsourcing Partners they have a wrong `outsourcingPartner_id` assigned. Maybe this might be the problem.

**Reply by Marvin-S** [reply]
I've created a PR #1156 so you can take a closer look at it. I've just set a break point to `repo.save(clonedData)` inside the belongs_to_many... test file and run through the program with the debugger.

**Reply by Marvin-S** [reply]
I finally figured out what is causing this issues 🎉
The problem is that Apollo transforms the data that it receives from the API to have single instances for the data. So you only have one instance of Billing Group 1, one instance of Billing Group 2 and so on. So Apollo realizes, that the Billing Group that is assigned to multiple Outsourcing Partners is the same and only creates one instance for it.

You can reproduce this behavior with the following data: 
```typescript
const billingGroups = [
      {
        id: 1,
        name: 'The Jedis',
        __typename: 'BillingGroup',
      },
      {
        id: 2,
        name: 'Jedi Council',
        __typename: 'BillingGroup',
      },
      {
        id: 3,
        name: 'Dark Forces',
        __typename: 'BillingGroup',
      },
    ]

    const outsourcingPartners = [
      {
        id: 1,
        name: 'Luke Skywalker',
        billingGroups: [billingGroups[0], billingGroups[1], billingGroups[2]],
        __typename: 'OutsourcingPartner',
      },
      {
        id: 2,
        name: 'Darth Vader',
        billingGroups: [billingGroups[0], billingGroups[1], billingGroups[2]],
        __typename: 'OutsourcingPartner',
      },
      {
        id: 3,
        name: 'Yoda',
        bbillingGroups: [billingGroups[0], billingGroups[1], billingGroups[2]],
        __typename: 'OutsourcingPartner',
      },
      {
        id: 4,
        name: 'Clone Trooper 0815',
        billingGroups: [billingGroups[0], billingGroups[1], billingGroups[2]],
        __typename: 'OutsourcingPartner',
      },
    ]
```

If you save the data like this, you can reproduce the problem:
```typescript
const repo = useRepo(OutsourcingPartner)
    repo.save(outsourcingPartners)
```

I've updated my PR, to remove dependencies from Apollo and GraphQL in general, so we can focus on the root of the problem :)

**Reply by CodeDredd** [reply]
Awesome work Marvin! Now i am going to look into a solution for this.

**Reply by CodeDredd** [reply]
Ok....i don't know if there is really an error in the normalizr. But i found a good solution.
Instead of using loadash `_clonedeep` use https://github.com/rhysd/fast-json-clone  or if you have `Date` objects https://github.com/davidmarkclements/rfdc. With that finally the refernces are removed.
Since we need a extensible object anyway, it seems to be the correct way.

The question maybe that i still have is if is should include this into the save function to check if the object is extensible and use this package.

**Reply by Marvin-S** [reply]
Yeah, I think you're right. I would say Normalizr works as designed in this case.

I will take a look at the rfdc solution, because we might have Date objects in the future.

I also thought about where this 'dereferencing' (if this is even a word 😄) should be placed: in Apollo, in the app code or inside pin-orm. I think one could argue for every option, but I think those references inside the object are not that uncommon, so I personally think that it would be a good place to implement this 'dereferencing' inside pinia-orm.

**Reply by CodeDredd** [reply]
Yeah i think i need an option for this. I am just thinking about if this is going to be an only graphql problem. Because in REST you will never have this problem.
